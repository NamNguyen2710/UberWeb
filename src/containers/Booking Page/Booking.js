import React from 'react';
import MyMap from './CurrentLocation';
import MapSearch from './MapSearch';
import BookingBoard from './BookingBoard';
import { Redirect } from 'react-router-dom';

class Booking extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      initSearch: [],
      from: '',
      to: '',
      redirect: null,
    }
    this.handleSearchFound = this.handleSearchFound.bind(this);
    this.revertToSearch = this.revertToSearch.bind(this);
    this.searchHome = this.searchHome.bind(this);
    this.confirmBooking = this.confirmBooking.bind(this);
  }

  handleSearchFound(info, value=null) {
    if (value) {
      this.setState({[value]: info});
    } else if (this.state.from === "") {
      this.setState({from: info});
    } else {
      this.setState({to: info});
    }
  }

  searchHome(value) {
    this.setState({initSearch: value});
  }

  revertToSearch(value) {

  }

  confirmBooking() {
    this.setState({ redirect: "/" })
  }

  render() {
    if (this.state.redirect !== null) return <Redirect to={this.state.redirect} />
    let bookingControl;
    if (this.state.from === '' || this.state.to === '') {
      bookingControl = 
        <MapSearch 
          handleSearchFound={this.handleSearchFound}
          initSearch={this.state.initSearch}
        />;
    } else { 
      bookingControl = 
        <BookingBoard 
          from={this.state.from}
          to={this.state.to}
          revertToSearch={this.revertToSearch}
          confirmBooking={this.confirmBooking}
        />
    }
    return (
      <div className="booking-page">
        <MyMap searchHome={this.searchHome}>
          {bookingControl}
        </MyMap>
      </div>
    );
  }
}

export default Booking;