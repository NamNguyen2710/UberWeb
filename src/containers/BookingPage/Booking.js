import React from "react";
import router from "../../router";
import MyMap from "./CurrentLocation";
import MapSearch from "./MapSearch";
import BookingBoard from "./BookingBoard";
import { Redirect } from "react-router-dom";
import { Marker, Popup } from "react-leaflet";

class Booking extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      initSearch: {},
      firstSearch: true,
      from: {
        position: null,
        info: "",
      },
      to: {
        position: null,
        info: "",
      },
      searchStage: "from",
      redirect: null,
    };
    this.handleSearchFound = this.handleSearchFound.bind(this);
    this.revertToSearch = this.revertToSearch.bind(this);
    this.searchHome = this.searchHome.bind(this);
    this.confirmBooking = this.confirmBooking.bind(this);
  }

  handleSearchFound(latlng, info) {
    if (
      this.state.initSearch.lat === latlng.lat &&
      this.state.initSearch.lng === latlng.lng
    ) {
      this.setState({
        [this.state.searchStage]: {
          position: latlng,
          info: info + " (Your Location)",
        },
      });
    } else {
      this.setState({
        [this.state.searchStage]: {
          position: latlng,
          info: info,
        },
      });
    }
    if (this.state.searchStage === "from")
      this.setState({
        searchStage: "to",
        firstSearch: false,
      });
  }

  searchHome(value) {
    this.setState({ initSearch: value });
  }

  revertToSearch(value) {
    this.setState({
      [value]: {
        position: null,
        info: "",
      },
      searchStage: value,
    });
  }

  confirmBooking() {
    this.setState({ redirect: router.HOME });
  }

  render() {
    if (this.state.redirect !== null)
      return <Redirect to={this.state.redirect} />;

    let bookingBoard;
    if (this.state.from.info !== "" && this.state.to.info !== "") {
      bookingBoard = (
        <BookingBoard
          from={this.state.from}
          to={this.state.to}
          revertToSearch={this.revertToSearch}
          confirmBooking={this.confirmBooking}
        />
      );
    } else {
      if (this.state.initSearch.hasOwnProperty("lat")) {
        bookingBoard = (
          <MapSearch
            handleSearchFound={this.handleSearchFound}
            initSearch={this.state.initSearch}
            firstSearch={this.state.firstSearch}
          />
        );
      }
    }

    return (
      <div className="booking-page">
        <MyMap searchHome={this.searchHome}>
          {bookingBoard}
          {this.state.from.position && this.state.from.position.lat && (
            <Marker position={this.state.from.position}>
              <Popup>{this.state.from.info}</Popup>
            </Marker>
          )}
          {this.state.to.position && this.state.to.position.lat && (
            <Marker position={this.state.to.position}>
              <Popup>{this.state.to.info}</Popup>
            </Marker>
          )}
        </MyMap>
      </div>
    );
  }
}

export default Booking;
