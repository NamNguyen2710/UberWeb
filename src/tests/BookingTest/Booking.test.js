import React from "react";
import Booking from "../../containers/BookingPage/Booking";
import MyMap from "../../containers/BookingPage/CurrentLocation";
import { mount } from "enzyme";
import { waitFor } from "@testing-library/react";
import toJson from "enzyme-to-json";
import { Marker } from "leaflet";

jest.mock(
  "../../containers/BookingPage/CurrentLocation",
  () => ({ children }) => children
);
jest.mock("../../containers/BookingPage/MapSearch", () => "map-search");
jest.mock("../../containers/BookingPage/BookingBoard", () => "booking-board");
jest.mock("react-leaflet", () => ({
  Marker: () => jest.fn(),
  Popup: () => jest.fn(),
}));

const compareObj = (obj1, obj2) =>
  JSON.stringify(obj1) === JSON.stringify(obj2);

const mockGeolocation = {
  getCurrentPosition: jest.fn(),
  watchPosition: jest.fn(),
};
global.navigator.geolocation = mockGeolocation;

describe("booking page", () => {
  it("render correctly", () => {
    const snapshot = mount(<Booking />);
    expect(toJson(snapshot)).toMatchSnapshot();
  });

  it("currentLocation found, search booking flow", async () => {
    const confirm = jest.fn();
    let state;

    const wrapper = mount(<Booking />);
    wrapper.instance().confirmBooking = confirm;

    //test: CurrentLocation return initLocation, set state.from to initLocation
    wrapper.instance().searchHome({ lat: 5, lng: 5 });
    wrapper.update();
    wrapper
      .find("map-search")
      .props()
      .handleSearchFound({ lat: 5, lng: 5 }, "");
    await waitFor(() => {
      wrapper.update();
      state = wrapper.instance().state;
      console.log(wrapper.debug());
      expect(wrapper.find("map-search").exists()).toBeTruthy();
      expect(
        !state.firstSearch &&
          state.searchStage === "to" &&
          compareObj(state.from.position, { lat: 5, lng: 5 }) &&
          state.from.info.includes("(Your Location)")
      ).toBeTruthy();
    });

    //test: MapSearch search state.to location
    wrapper
      .find("map-search")
      .props()
      .handleSearchFound({ lat: 10, lng: 15 }, "Work Location");
    await waitFor(() => {
      wrapper.update();
      state = wrapper.instance().state;
      expect(wrapper.find("map-search").exists()).toBeFalsy();
      expect(wrapper.find("booking-board").exists()).toBeTruthy();
      expect(
        compareObj(state.to.position, { lat: 10, lng: 15 }) &&
          state.searchStage === "to"
      ).toBeTruthy();
    });

    //test: BookingBoard revert to search state.from location
    wrapper.find("booking-board").props().revertToSearch("from");
    await waitFor(() => {
      wrapper.update();
      state = wrapper.instance().state;
      expect(
        state.from.position === null &&
          state.to.position !== null &&
          state.searchStage === "from"
      ).toBeTruthy();
      expect(wrapper.find("booking-board").exists()).toBeFalsy();
      expect(wrapper.find("map-search").exists()).toBeTruthy();
    });

    //test:search state.from location
    wrapper
      .find("map-search")
      .props()
      .handleSearchFound({ lat: 12, lng: 63 }, "Home Location");
    await waitFor(() => {
      wrapper.update();
      state = wrapper.instance().state;
      expect(
        compareObj(state.from.position, { lat: 12, lng: 63 }) &&
          state.searchStage === "to"
      ).toBeTruthy();
      expect(wrapper.find("map-search").exists()).toBeFalsy();
      expect(wrapper.find("booking-board").exists()).toBeTruthy();
    });

    //test: revert to search state.to
    wrapper.find("booking-board").props().revertToSearch("to");
    await waitFor(() => {
      wrapper.update();
      state = wrapper.instance().state;
      expect(
        state.to.position === null &&
          state.from.position !== null &&
          state.searchStage === "to"
      ).toBeTruthy();
      expect(wrapper.find("booking-board").exists()).toBeFalsy();
      expect(wrapper.find("map-search").exists()).toBeTruthy();
    });

    //test: submit
    wrapper
      .find("map-search")
      .props()
      .handleSearchFound({ lat: 10, lng: 15 }, "Work Location");
    wrapper.update();
    wrapper.find("booking-board").props().confirmBooking();
    await waitFor(() => {
      expect(confirm).toHaveBeenCalled();
    });
  });
});
