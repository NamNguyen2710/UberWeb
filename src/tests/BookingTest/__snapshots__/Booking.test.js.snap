// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`booking page render correctly 1`] = `
<Booking>
  <div
    className="booking-page"
  >
    <Component
      searchHome={[Function]}
    />
  </div>
</Booking>
`;
